src/ZVal.h

```diff
@@ -0,0 +1,505 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+// Values used in ZAM execution, and also for representing records and
+// vectors during interpreter execution.
+
+#pragma once
+
+#include <unordered_set>
+
+#include "Dict.h"
+#include "Expr.h"
+#include "IntrusivePtr.h"
+
+ZEEK_FORWARD_DECLARE_NAMESPACED(StringVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(AddrVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(SubNetVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(File, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(Func, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(ListVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(OpaqueVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(PatternVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(TableVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(RecordVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(VectorVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(Type, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(Val, zeek);
+
+namespace zeek {
+
+// Only needed for compiled code.
+class IterInfo;
+
+typedef std::vector<zeek::ValPtr> val_vec;
+
+// A bit of this mirrors BroValUnion, but BroValUnion captures low-level
+// representation whereas we aim to keep Val structure for more complex
+// Val's.
+//
+// Ideally we'd use IntrusivePtr's for memory management, but we can't
+// given we have a union and thus on destruction C++ doesn't know which
+// member flavor to destruct.
+//
+// Note that a ZAMValUnion by itself is ambiguous: it doesn't track its
+// type.  This makes them consume less memory and cheaper to copy.  It
+// does however require a separate way to determine the type.  Generally
+// this is doable using surrounding context, or can be statically determined
+// in the case of optimization/compilation.
+//
+// An alternative would be to use std::variant, but (1) it tracks the
+// variant type, and (2) it won't allow access to the managed_val member,
+// which not only simplifies memory management but also is required for
+// sharing of ZAM frame slots.
+union ZAMValUnion {
```

Comment:
Given that we're merging this outside of ZAM into the main part of the code, I'd prefer to rename the "ZAM" types (`ZAMValUnion`, `ZAM_record`, etc) to something independent of ZAM. Otherwise I can see things getting confusing because the ZAM terminology seems to suggest that it's something specific to the optimizer.

Reply:
How about `ZValUnion` / `Z_vector` / `Z_record` ?

Reply:
(with `Z` standing for `Zeek`)

Reply:
Hmm, not sure. Then we'd end up with having both `BroValUnion` and `Z(eek)ValUnion`. While I don't have a better idea right now, maybe this is actually hinting as something else: there's a lot of overlap between new and existing types. Is there any chance we could unify that so that we end up with just single representation that covers all use cases? I assume the answer is no (otherwise you'd have done it), but just seeing the current code leaves me wondering if the structure is quite right yet. (I'm not deep enough in the code to suggest something else right now)

Reply:
Here's the core problem as I see it.  Memory management for aggregate values is handled by `Val` reference-counting rather than direct reference-counting.  For example, a `RecordVal`'s low-level representation in `BroValUnion` is `PDict<TableEntryVal>*`, and dictionaries do not support reference-counting.  Thus, to share an aggregate (as often required by Zeek scripting semantics), the sharing has to be done using the `RecordVal` rather than the low-level representation.

One approach would be to change all of these low-level representations to be `Obj`'s, though getting this right w/o introducing leaks would likely be a pain.  If we made that change, then `ZAMValUnion` could be expressed as some additional fields to the existing (modified) `BroValUnion` rather than a separate representation.

Another approach is to get rid of `BroValUnion` and hoist its different variants into the different `Val` subclasses that use them.  For example, `RecordVal`'s would have the `PDict` as an explicit member variable, rather than using the common `BroValUnion` to store them.  The scope of this change also gives me pause, though, and it could wind up being clunky or potentially buggy for closely-related `Val` subclasses like `TimeVal` and `DoubleVal`.  If we did this then we'd still have `ZAMValUnion` but it would be the only such low-level representation of values (other than those spread across different `Val` subclasses).